src/goto-instrument/code_contracts.cpp

```diff
@@ -1809,3 +1875,277 @@ exprt assigns_clauset::compatible_expression(
+  }
+
+  // fixing the function name.
+  to_symbol_expr(call.function()).set_identifier(fn_name);
+  log.debug() << "printing updated call expression: "
+              << expr2c(call, parent.get_namespace()) << "\n";
+
+  ins->set_function_call(call);
+}
+
+/* Declarations for contract enforcement */
+
+is_fresh_enforcet::is_fresh_enforcet(
+  code_contractst &_parent,
+  messaget _log,
+  irep_idt _fun_id)
+  : is_fresh_baset(_parent, _log, _fun_id)
+{
+  std::stringstream ssreq, ssensure, ssmemmap;
+  ssreq << CPROVER_PREFIX << fun_id << "_requires_is_fresh";
+  this->requires_fn_name = ssreq.str();
+
+  ssensure << CPROVER_PREFIX << fun_id << "_ensures_is_fresh";
+  this->ensures_fn_name = ssensure.str();
+
+  ssmemmap << CPROVER_PREFIX << fun_id << "_memory_map";
+  this->memmap_name = ssmemmap.str();
+}
+
+void is_fresh_enforcet::create_declarations()
+{
+  std::ostringstream oss;
+  std::string cprover_prefix(CPROVER_PREFIX);
+  oss << "static _Bool " << memmap_name
+      << "[" + cprover_prefix + "constant_infinity_uint]; \n"
+      << "\n"
+      << "_Bool " << requires_fn_name
+      << "(void **elem, " + cprover_prefix + "size_t size) { \n"
+      << "   *elem = malloc(size); \n"
+      << "   if (!*elem || " << memmap_name
+      << "[" + cprover_prefix + "POINTER_OBJECT(*elem)]) return 0; \n"
+      << "   " << memmap_name << "[" + cprover_prefix
+      << "POINTER_OBJECT(*elem)] = 1; \n"
+      << "   return 1; \n"
+      << "} \n"
+      << "\n"
+      << "_Bool " << ensures_fn_name
+      << "(void *elem, " + cprover_prefix + "size_t size) { \n"
+      << "   _Bool ok = (!" << memmap_name
+      << "[" + cprover_prefix + "POINTER_OBJECT(elem)] && "
+      << cprover_prefix + "r_ok(elem, size)); \n"
+      << "   " << memmap_name << "[" + cprover_prefix
+      << "POINTER_OBJECT(elem)] = 1; \n"
+      << "   return ok; \n"
+      << "}";
+
+  add_declarations(oss.str());
+}
+
+void is_fresh_enforcet::create_requires_fn_call(goto_programt::targett &ins)
+{
+  update_fn_call(ins, requires_fn_name, true);
+}
+
+void is_fresh_enforcet::create_ensures_fn_call(goto_programt::targett &ins)
+{
+  update_fn_call(ins, ensures_fn_name, false);
+}
+
+/* Declarations for contract replacement: note that there may be several
+   instances of the same function called in a particular context, so care must be taken
+   that the 'call' functions and global data structure are unique for each instance.
+   This is why we check that the symbols are unique for each such declaration.  */
+
+std::string unique_symbol(const symbol_tablet &tbl, const std::string &original)
+{
+  auto size = tbl.next_unused_suffix(original);
+  return original + std::to_string(size);
+}
+
+is_fresh_replacet::is_fresh_replacet(
+  code_contractst &_parent,
+  messaget _log,
+  irep_idt _fun_id)
+  : is_fresh_baset(_parent, _log, _fun_id)
+{
+  std::stringstream ssreq, ssensure, ssmemmap;
+  ssreq /* << CPROVER_PREFIX */ << fun_id << "_call_requires_is_fresh";
```

Comment:
Are these commented out CPROVER_PREFIX's deliberate? is_fresh_enforcet::is_fresh_enforcet seems to insert the, but they are commented out here - just wondering which is correct, or whether both are?